[CORE-467] Copy compact entities

samanehsan opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/CORE-467

Changes:
- In `entityService.copyEntities`, check that both the source and destination workspace have the same setting for CompactDataTablesSetting. If it is only enabled for one, throw an exception
- Implement copyEntities in CompactEntityProvider:
  - If the destination workspace already contains a record for any of the entities being copied, the request fails (these are hard conflicts)
  - Otherwise, check if the destination workspace contains a record for any of the downstream references of the entities being copied (soft conflicts). If there are no soft conflicts or `linkExistingEntities` is true, copy the entities and exclude any soft conflicts. If `linkExistingEntities` is false, do not copy anything and return the soft conflicts.

The legacy copyEntities code has different behavior for checking soft conflicts. It recursively builds a path of references for each entity being copied. If the destination workspace contains the last entity in the path it is considered a soft conflict. With the new implementation, if ANY of the references of the entity exist for the destination workspace, it is considered a soft conflict. Because we do not need to know the entity path, this also changes the response when the request fails due to soft conflicts (and linkExistingEntities is false). Instead of EntitySoftConflict including the conflicts recursively, it will be flattened: `EntitySoftConflict(entityToCopy.entityType, entityToCopy.entityName, Seq(EntitySoftConflict(entityRef.entityType, entityRef.entityName, Seq()), EntitySoftConflict(entityRef2.entityType, entityRef2.entityName, Seq())...))`

---

**PR checklist**

- [x] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should publish a new official `rawls-model` and perform the corresponding dependency updates as specified in the [README](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model):
  - [ ] in the automation subdirectory
  - [ ] in workbench-libs
  - [ ] in firecloud-orchestration
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

samanehsan committed:
Copy compact entities

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -208,6 +201,87 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+      getReferencesTo(workspaceId, Seq(entity)).map { references =>
+        references.map(ref => (entity, ref))
+      }
+    }
```
I think you should figure out how to do this in a single query

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -208,6 +201,87 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+    } yield allRefs
+  }
+
+  def recursiveGetEntityReferences(workspaceId: UUID,
```
I think you should try a recursive query https://dev.mysql.com/doc/refman/9.0/en/with.html#common-table-expressions-recursive

dvoet replied:
you could try using AI with a prompt like "reimplement recursiveGetEntityReferences using a recursive CTE"

davidangb replied:
+1 to using a recursive CTE; I think it should be feasible against the `ENTITY_REFS` table

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -208,6 +201,87 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+        reduceSqlActionsWithDelim(typeNameClauses.toSeq, sql" or "))
+             sql""" );"""
+      sql.asUpdate
+  }
```
this looks correct for the `ENTITY` table. You'll need a second `insert into … select` statement to populate the `ENTITY_REFS` table for the `destWorkspaceId`, too.

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -208,6 +201,87 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+        entityRefs.grouped(driverComponent.batchSize)
+      } else {
+        Iterator(entityRefs)
+      }
```
I don't think you need the if statement, `entityRefs.grouped(driverComponent.batchSize)` will return only 1 group if the size is less than the batch size

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -208,6 +201,87 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+        Iterator(entityRefs)
+      }
+
+    val allCopies = DBIO.sequence(chunks map copyChunkOfEntitiesOrAllEntities)
```
can you replace `copyChunkOfEntitiesOrAllEntities` with  `copyEntities(sourceWs, destWs, _)`

samanehsan committed:
Code review comments

- Changed soft conflict behavior to return
  a soft conflict if any downstream references
  already exist in the destination workspace
  (instead of the last one in the path)

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-467-copy-entities

samanehsan committed:
Use EntityPointer instead of AttributeEntityReference

samanehsan committed:
Add test for successful copyEntities

samanehsan committed:
Add copyEntity test with hard conflict

samanehsan committed:
Add copyEntity test for soft conflict and linkExistingEntities = false

samanehsan committed:
Scala fmt

samanehsan committed:
Add copyEntity test when soft conflicts and linkExistingEntities = true

samanehsan committed:
Only copy entities between workspaces with the same data table settings

Either both workspaces must have compact data tables enabled, or both
must have the setting disabled.

samanehsan committed:
Fix scala fmt

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -223,6 +224,129 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+    val refs = entityNames.map(name => EntityPointer(entityType, name))
+    for {
+      startingEntityRecords <- getEntityRefs(workspaceId, refs)
+      entities = startingEntityRecords.map(record => record.toPointer)
```
This performs an extraneous database lookup in `getEntityRefs` which looks totally unnecessary. Annotated with comments:
```scala
    // translate the entityType/entityName arguments to a Set[EntityPointer]
    val refs = entityNames.map(name => EntityPointer(entityType, name))
    for {
      // look up each pointer in the database and return as CompactEntityRecords
      startingEntityRecords <- getEntityRefs(workspaceId, refs)
      // translate the CompactEntityRecords back to pointers
      // this Seq has the exact same contents as the `refs` Set above
      entities = startingEntityRecords.map(record => record.toPointer)
```
unless you are using this to validate that all the input entityType/entityName values exist? If that's the goal, look at the `existsAll` function instead, which should be much more efficient.

samanehsan replied:
I think I needed CompactEntityRecords in a previous iteration but I agree it doesn't make sense now! I'll just use the refs and skip the db look up

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -223,6 +224,129 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+
+  def recursiveGetEntityReferences(workspaceId: UUID,
+                                   entities: Set[EntityPointer]
+                                  ): ReadAction[Map[EntityPointer, Set[EntityPointer]]] = {
```
Consider using the `RefMapping` case class instead of `EntityPointer, Set[EntityPointer]`

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -223,6 +224,129 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+        from EntityReferences
+      """
+
+      concatSqlActions(baseSql, recursiveSql, finalSql).as[(String, String, String, String)].map { rows =>
```
We don't have a pre-existing case class that's better to use than `(String, String, String, String)`. As I went searching I realized that the `RefPointerRecord` _should_ be what to use here, but it is incorrect since #3305; I should have updated it in that PR. Perhaps update that case class and use it here?

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -113,7 +116,91 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+      // return the entities already present in the destination workspace
+      repository.dataSource.inTransaction { _ =>
+        repository.queries.getEntityRefs(destWorkspaceContext.workspaceIdAsUUID, paths).map { conflicts =>
+          val conflictsAsRefs = conflicts.toSeq.map(_.toPointer)
+          paths.filter(p => conflictsAsRefs.contains(p))
```
can this just return `conflicts`? Does `paths.filter(p => conflictsAsRefs.contains(p))` add any functionality here?

samanehsan replied:
Good catch! This is no longer needed after using `Set[EntityPointer]` instead of of `Seq[EntityPath]`

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -113,7 +116,91 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+                                                                     destWorkspaceContext.workspaceIdAsUUID,
+                                                                     entitiesToCopy
+                  )
+                  _ <- repository.updateLastModified(workspaceId)
```
check out some of the other methods in this provider, like `batchUpsertEntities`; they don't update the workspace last-modified date in the same transaction. Could you adopt that?

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -223,6 +224,129 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+        entityRefsCopiedCount <- copyEntityReferences(sourceWs, destWs, chunk)
+      } yield (entitiesCopiedCount, entityRefsCopiedCount)
+
+    val chunks: Iterator[Set[EntityPointer]] = entityRefs.grouped(driverComponent.batchSize)
```
I'm not sure that `driverComponent.batchSize` is going to be the optimal chunk size; that setting was chosen for the legacy implementation.

The only reason I see to chunk here is to ensure that the where clause that checks entityType/entityName is not outrageous. `driverComponent.batchSize` is set to 2000 [here](https://github.com/broadinstitute/terra-helmfile/blob/64cf867991757a9ad30e3244862576c7733ae8b5/charts/rawls/templates/config/_conf.tpl#L201) and I bet this code could use something larger. Perhaps add a new value to CompactEntityProviderConfig ?

davidangb reviewed (commented):
I did not review the test code. I have some specific comments inline, but I think this is on track! Thank you :)

samanehsan committed:
Fix failing tests

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -113,7 +116,89 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+                           parentContext: RawlsRequestContext
+  ): Future[EntityCopyResponse] = {
+
+    def getSoftConflicts(paths: Set[EntityPointer]): Future[Set[EntityPointer]] =
```
`paths` isn't a good name anymore

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -113,7 +116,89 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+      parentContext
+    )
+
+  def checkAndCopyEntities(sourceWorkspaceContext: Workspace,
```
this function is not used anywhere else, I would inline it

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -113,7 +116,89 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+
+    def getSoftConflicts(paths: Set[EntityPointer]): Future[Set[EntityPointer]] =
+      // return the entities already present in the destination workspace
+      repository.dataSource.inTransaction { _ =>
```
Does it make sense to have 1 transaction for this whole function instead of calling `inTransaction` 4 times? This would make immune from race conditions where data is updated between transactions but perhaps introduce locking issues.

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -113,7 +116,89 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+          pathsAndConflicts.flatMap { case (entityReferenceMap, entityReferences, softConflicts) =>
+            if (softConflicts.isEmpty || linkExistingEntities) {
+              val allEntityRefs: Set[EntityPointer] = entityReferenceMap.keys.toSet concat entityReferences
+              val allConflictRefs: Set[EntityPointer] = softConflicts
```
inline

samanehsan committed:
Code review comments

samanehsan committed:
Use one transaction for copyEntities

samanehsan committed:
Add CompactEntityProviderConfig batchCopyBatchSize

samanehsan committed:
Fix RefPointerRecord case class

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-467-copy-entities

samanehsan committed:
Fix tests

samanehsan committed:
Revert adding CompactEntityProviderConfig batchCopyBatchSize

Requires changing several places in the code and breaks a bunch
of tests

samanehsan committed:
Fix tests

samanehsan committed:
Fix compilation error

samanehsan committed:
Split up copyEntities function

samanehsan committed:
Fix test

samanehsan committed:
Use a custom batch size for copying entities

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -76,6 +77,17 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+        r.<<
+      )
+    )
+
```
scalafmt is weird sometimes.

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -114,7 +117,89 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+  ) = {
+    val allEntityRefs: Set[EntityPointer] = entityReferenceMap.map(_.from) ++ entityReferences
+    val entitiesToCopy: Set[EntityPointer] = allEntityRefs diff softConflicts
+    repository.queries
+      .copyEntitiesToNewWorkspace(
+        sourceWorkspaceId,
+        destWorkspaceId,
+        entitiesToCopy,
+        config.batchCopyBatchSize
+      )
+    EntityCopyResponse(
+      entitiesToCopy.map(_.toAttributeEntityReference).toSeq,
+      Seq.empty,
+      Seq.empty
+    )
```
Does this actually execute the `copyEntitiesToNewWorkspace` query? It looks like the query might be bypassed and the `EntityCopyResponse` object returned. You might need to `map` the query to the `EntityCopyResponse`, like:
```scala
repository.queries
      .copyEntitiesToNewWorkspace(
        sourceWorkspaceId,
        destWorkspaceId,
        entitiesToCopy,
        config.batchCopyBatchSize
      ) map { _ =>
      EntityCopyResponse(
        entitiesToCopy.map(_.toAttributeEntityReference).toSeq,
        Seq.empty,
        Seq.empty
      )
    }
```
I see your tests in `CompactEntityProviderSpec` which verify this gets called, which confuse me. Maybe this is called in a fire-and-forget mode and we might not wait for it? I am worried about what this is actually doing.

dvoet replied:
@davidangb is correct. The tests may verify the query function is called, but the result is a DBIO action which is never executed.

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProviderConfig.scala:
```diff
@@ -2,5 +2,6 @@ package org.broadinstitute.dsde.rawls.entities.compact
   // number of entities to handle in a single SQL statement when processing batchUpsert/batchUpdate
-  batchUpsertBatchSize: Int = 2500
+  batchUpsertBatchSize: Int = 250,
+  batchCopyBatchSize: Int = 4000
```
did you mean to change the `batchUpsertBatchSize` value?

samanehsan replied:
Oh! That was a mistake

davidangb reviewed (commented):
Some comments inline - in addition to those:
1. Could you add some scaladoc comments to the new runtime methods to explain what they do?
2. Could you add some tests to CompactEntityComponentSpec (for individual CompactEntityComponent methods) or CompactEntityProviderE2ESpec which verify functionality using the database?

davidangb commented:
Oh, one other question: have you looked at execution plans and/or assessed performance, especially at different batch sizes?

samanehsan commented:
> Oh, one other question: have you looked at execution plans and/or assessed performance, especially at different batch sizes?

I have not! Do you have any suggestions for how to do that?

samanehsan committed:
Code review comments

- Fix copying entities call
- Fix typo in entity provider config

samanehsan committed:
More code review changes

- Add docstrings
- Remove getEntitySubtrees and call recursiveGetEntityReferences
  directly

samanehsan commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -223,6 +235,164 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
     }
 
+  /**
+   * Copies entities and their references from a source workspace to a destination workspace.
```
I asked copilot to write these docstrings. Listing the params and return value felt too verbose so I deleted that part. What do you think?

samanehsan committed:
Add tests to CompactEntityComponentSpec

samanehsan committed:
Add copy entity E2E test

samanehsan committed:
Do not copy entities that do not exist in source workspace

samanehsan committed:
Add test for getting entity refs with cycle

samanehsan committed:
For testing different batch sizes

samanehsan commented:
I tested out copying ~5k and ~10k entities with different batch sizes using dev workspaces and running this Rawls branch locally. I used one entity with a lot of references to make the API call easier (just needed to list one entity name instead of 5k or 10k). What do you think of using 5000 as the batch size @davidangb? 

**BatchSize 2500**
1 entity, 5000 refs took 5600ms
1 entity, 10000 refs took 6399ms

**BatchSize 3000**
1 entity, 5000 refs took 4754ms
1 entity, 10000 refs 5794ms

**BatchSize 4000**
1 entity, 5000 refs took 4423ms
1 entity, 10000 refs took 5342ms

**BatchSize 5000**
1 entity, 5000 refs took 3902ms
1 entity, 10000 refs took  5372ms

samanehsan committed:
Fix unit tests

samanehsan committed:
Undo logging from testing

davidangb commented:
> What do you think of using 5000 as the batch size @davidangb?

Seems totally reasonable to me, your tests bear that out!

samanehsan committed:
Bump up batchCopyBatchSize

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-467-copy-entities

samanehsan committed:
Fix scalafmt

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -227,6 +239,172 @@ class CompactEntityQuery(driverComponent: DriverComponent)
+      val typeNameClauses = generateTypeNameSql(refs)
+      val sql = concatSqlActions(
+        sql"""insert into ENTITY(name, entity_type, workspace_id, record_version, deleted, attributes)
+             select name, entity_type, $destWorkspaceId, record_version, 0, attributes
```
nit: should the new entities in the destination workspace maintain the current record_version from the source workspace, or should we consider them brand new and always give them a record_version of 0? What does the legacy implementation do?

This isn't super important - either approach is fine and won't impact end users; this is more about semantics.

dvoet replied:
I don't think it matters. The only semantics we care about is that record_version is monotonically increasing.

samanehsan replied:
The legacy implementation keeps the same record version from the source workspace: https://github.com/broadinstitute/rawls/blob/cf74c035742140dae4f3b9fecb313a16a35d2094/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala#L747

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala:
```diff
@@ -227,6 +239,172 @@ class CompactEntityQuery(driverComponent: DriverComponent)
+    } else if (entities.size > batchSize) {
+      val batches = entities.grouped(batchSize).toSeq
+      DBIO
+        .sequence(batches.map(batch => recursiveGetEntityReferences(workspaceId, batch.toSet)))
```
nit: `batch.toSet` is redundant; `batch` is already a Set

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala:
```diff
@@ -114,7 +117,101 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments,
+                val entityReferences = entityReferenceMap.flatMap(_.to)
+                repository.queries.getEntityRefs(destWorkspaceContext.workspaceIdAsUUID, entityReferences).flatMap {
+                  conflicts =>
+                    val softConflicts = conflicts.toSeq.map(_.toPointer).toSet
```
nit: `conflicts.toSeq` is redundant; `conflicts` is already a Seq

davidangb commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponentSpec.scala:
```diff
@@ -2122,6 +2122,116 @@ class CompactEntityComponentSpec extends TestDriverComponentWithFlatSpecAndMatch
+    val entity3 =
+      Entity("entityName3", "entityType2", Map(AttributeName.withDefaultNS("attr3") -> AttributeString("value3")))
+
+    // Insert entity references between entities
+    val refMapping = RefMapping(entity1.toPointer, Set(entity2.toPointer, entity3.toPointer))
+    runAndWait(q.insertReferences(sourceWorkspaceId, Set(refMapping)))
```
no changes requested for this PR. Once the new serialization format merges, this test will need to be rewritten.

davidangb reviewed (commented):
Some very minor comments/nits inline; overall looks good!

For the execution plans, could you check the actual plans from the db? Our doc on how to do that is at https://broadworkbench.atlassian.net/wiki/spaces/CORE/pages/3785588737/Execution+Plans.

In most cases I am guessing the SQL queries will be fine, but I still think it's worth checking the plans.

samanehsan committed:
Add comments about query execution plan

samanehsan commented:
This was the result of explaining the recursive CTE. Do you think the comment includes all the relevant details?
```
*************************** 1. row ***************************
           id: 1
  select_type: PRIMARY
        table: <derived2>
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 4
     filtered: 100.00
        Extra: NULL
*************************** 2. row ***************************
           id: 2
  select_type: DERIVED
        table: er
   partitions: NULL
         type: ref
possible_keys: unq_from_to,idx_to
          key: idx_to
      key_len: 16
          ref: const
         rows: 1
     filtered: 5.00
        Extra: Using where; Using index
*************************** 3. row ***************************
           id: 3
  select_type: UNION
        table: er1
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 2
     filtered: 100.00
        Extra: Recursive
*************************** 4. row ***************************
           id: 3
  select_type: UNION
        table: er
   partitions: NULL
         type: ref
possible_keys: unq_from_to,idx_to
          key: idx_to
      key_len: 16
          ref: const
         rows: 1
     filtered: 5.00
        Extra: Using where; Using index
*************************** 5. row ***************************
           id: 5
  select_type: UNION RESULT
        table: <union2,3>
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: NULL
     filtered: NULL
        Extra: Using temporary
5 rows in set, 1 warning (0.12 sec)

```

davidangb reviewed (approved):
thank you for the execution plans, looks good! I do think that plan for the recursive queries explains it.
